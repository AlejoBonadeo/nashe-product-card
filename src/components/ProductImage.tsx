import React from "react";
import { useContext } from "react";
import { ProductContext } from "./ProductCard";
import noImage from "../assets/no-image.jpg";
import styles from "../styles/styles.module.css";

export const ProductImage = ({ img = "", className= "" }) => {
  const { product } = useContext(ProductContext);
  const imgToShow = img || product.img || noImage;
  return <img className={`${styles.productImg} ${ className }`} src={imgToShow} alt="Product" />;
};

