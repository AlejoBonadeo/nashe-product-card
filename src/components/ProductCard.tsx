import React from "react";
import styles from "../styles/styles.module.css";
import useProduct from "../hooks/useProduct";
import { createContext } from "react";
import { OnChangeArgs, Product, ProductContextProps, InitialState, ProductCardHandlers } from '../interfaces/interfaces';
import {ProductImage} from "./ProductImage";
import {ProductTitle} from "./ProductTitle";
import {ProductButtons} from "./ProductButtons";

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

interface CardProps {
  product: Product;
  children: ( args: ProductCardHandlers ) => JSX.Element
  className?: string;
  style?: React.CSSProperties;
  onChange?: ( args: OnChangeArgs ) => void;
  value?: number
  initialState?: InitialState
}

const ProductCard = ({ children, product, className, style, onChange, value, initialState }: CardProps) => {

  const { counter, increaseBy, isMaxCountReached, reset } = useProduct( { onChange, product, value, initialState } );

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount: initialState?.maxCount
      }}
    > 
      <div className={`${styles.productCard} ${ className }`} style={style}>
        {
          children({
            count: counter,
            isMaxCountReached,
            maxCount: initialState?.maxCount,
            product,
            reset,
            increaseBy
          })
        }
      </div>
    </Provider>
  );
};

ProductCard.Title = ProductTitle;
ProductCard.Image = ProductImage;
ProductCard.Buttons = ProductButtons;

export default ProductCard;
