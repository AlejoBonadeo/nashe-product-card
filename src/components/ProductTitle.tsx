import React from "react";
import { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from "../styles/styles.module.css";

type TitleProps = {
  title?: string
  className?: string
}

export const ProductTitle = ({ title, className }: TitleProps) => {
  const { product } = useContext(ProductContext);
  return (
    <span className={`${styles.productDescription} ${ className }`}>
      {" "}
      {title || product.title}{" "}
    </span>
  );
};
