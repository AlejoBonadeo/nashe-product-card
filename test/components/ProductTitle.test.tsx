import * as React from 'react';
import renderer from 'react-test-renderer'
import { ProductTitle } from '../../src/components/ProductTitle';

describe('ProductTitle', () => {
    test('debe de mostrar el componente correctamente en el titulo personalizado', () => {
        const wrapper = renderer.create(
            <ProductTitle title="Custom Product"/>
        )
        expect(wrapper.toJSON()).toMatchSnapshot()
    })
    test('debe de mostrar el componente con el nombre del producto', () => {
        
    })
    
})