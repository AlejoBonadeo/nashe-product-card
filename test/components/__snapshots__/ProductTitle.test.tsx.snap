// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe de mostrar el componente correctamente en el titulo personalizado 1`] = `
<span
  className="productDescription undefined"
>
   
  Custom Product
   
</span>
`;
